feat: add locustfile for API load testing

ianwestcott opened the pull request:
Asana Task: [💳 ECS application prototype part 2: auto-scaling](https://app.asana.com/0/1113179098808463/1198556978865359/f)

**WIP**

Adds [Locust](https://locust.io/) configuration for load-testing the API application.

Usage: 

```
$ LOCUST_API_KEY="<api-key>" locust --host="https://api-url"
```

TODO:

- Investigate intermittent failing requests on dev-blue
- Maybe: Adjust frequency of tasks to reduce frequency of those that hit more expensive endpoints/increase frequency for lower-overhead endpoints?
- Confirm that ECS autoscaling behavior matches our expectations relative to existing Elastic Beanstalk environments

ianwestcott commented:
Next steps:
1. Clean up this PR:
   - remove `for` loops, pick a random item from each list instead
   - Declare Python requirements for running load test (poetry?)
1. Update api-dev-blue (ECS) and api-dev (Beanstalk) to match api-prod:
   - Same instance type (or equivalent CPU/memory)
   - Same autoscaling configuration (+3,-1)
1. Build dashboards for real-time monitoring of api-dev CPU/Memory/instance count
1. Run load tests against api-dev-blue and api-dev and compare the results
1. Figure out where the Locustfile should live long term?

paulswartz commented on .tool-versions:
```diff
@@ -1,2 +1,3 @@
 elixir 1.10.3-otp-23
 erlang 23.0.2
+python 3.9.1
```
Can we bump this to `3.9.2` since that's the latest version, and what `gtfs_creator` is using?

ianwestcott replied:
Given that a month has passed since this comment, and that `3.9.3` contained some security fixes, I'm going to jump straight to `3.9.4` if that's alright. https://www.python.org/downloads/release/python-394/

paulswartz replied:
https://www.python.org/downloads/release/python-395/

paulswartz commented on locustfile.py:
```diff
@@ -0,0 +1,225 @@
+        if "name" in kwargs.keys():
+            name = kwargs["name"]
+        else:
+            name = None
```
`name = kwargs.get("name")` but really what you want is to have this provided as default keyword arguments.

```python
def api_request(self, endpoint, name=None, filters=None, fields=None, include=None, sort=None):
```

paulswartz commented on locustfile.py:
```diff
@@ -0,0 +1,225 @@
+                    fltr_val = ",".join(filters[fltr])
+                else:
+                    fltr_val = str(filters[fltr])
+                params[f"filter[{fltr}]"] = fltr_val
```
```python
if filters is not None:
  for fltr, fltr_val in filters.items():
    if isinstance(fltr_val, list):
      fltr_val = ",".join(values)
    params[f"filter[{fltr}]"] = fltr_val
```

paulswartz commented on locustfile.py:
```diff
@@ -0,0 +1,225 @@
+                             include=["stop", "trip", "prediction"],
+                             sort="arrival_time")
+
+    @task(len(heavy_rail_routes))
```
This seems like it'll run N*N API calls for the heavy rail routes? Or does `@task(count)` not run the given task `count` times?

paulswartz reviewed (approved):
I tried this out, it seems to work!

ianwestcott commented:
Small updates:
- Bumped Python version to 3.9.5
- Moved the locustfile to a `load_tests` subdirectory (matching the convention established in [alerts_concierge](https://github.com/mbta/alerts_concierge/tree/master/load_tests)
- Added a requirements.txt file

losvedir commented:
I've come around on the idea of including this in the repo itself. Seems like a useful tool to have at hand for testing things locally, too. With that said then, can you tweak it a little to make it consistent with the TRC team's python practices? Those things are:

1) We use `poetry` to manage dependencies. Can you add the latest `poetry` to `.tool-versions`.
2) Instead of `requirements.txt`, please do a `poetry init` to use `pyproject.toml`
3) In `pyproject.toml` add `locust`, `flake8`, and `black`
4) Do `poetry run flake8` and `poetry run black` to make sure the code uses our familiar style
5) Update the README where applicable (at least the python/poetry asdf plug-ins).

Thanks!

ianwestcott commented:
@losvedir I made changes based on those recommendations. Let me know if I should update and/or squash this branch.

losvedir reviewed (approved):
Looks good, thanks! Can you please squash and merge?
